### ISSUE TITLE
Parallel make of "tests" target fails

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#2354]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/2350
- Created at 2014-06-04 11:09:55 by [nkondras](https://pagure.io/user/nkondras)
- Closed at 2020-03-24 14:17:56 as wontfix
- Assigned to nobody
---

Execution of "make -j4 tests" is seen failing on Fedora 20 VM with probability of about 65%.

Failures seem to be due to race conditions, such as some files built in two processes at the same time. The problem might be due to a .PHONY target used somewhere in the dependency tree.

A similar-effect command "make -j4 check LOG_COMPILER=true" completes reliably without failures.

### Comments

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 15:52:11](https://pagure.io/SSSD/sssd/issue/2350#226525)**

Log file from build will be helpful. There should not be problem with building some file in two processes at the same time. libtool should prevent this problem.

It will generate prefix for object files
eg.

       gcc -o src/tools/sss_cache-tools_util.o src/tools/tools_util.c
                        ^^^^^^^^^ 

cc:  => lslebodn@redhat.com

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 16:37:25](https://pagure.io/SSSD/sssd/issue/2350#226526)**

attachment
[make_tests_logs.tar.xz](/SSSD/sssd/issue/raw/files/14a3c91b093d9c0f1c59ae36c7f43e216654d96a4794c019d7769f525bfd36a3-make_tests_logs.tar.xz)

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 16:39:30](https://pagure.io/SSSD/sssd/issue/2350#226527)**

Attached logs of 20 consecutive runs of "make -j4 tests; make -j4 clean" on an otherwise unloaded Fedora 20 VM. Of these 13 have failed. You can find the failed ones by grepping for "Error" among them.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 16:41:11](https://pagure.io/SSSD/sssd/issue/2350#226528)**

The failures are mostly different. They reproduce with make debugging enabled ("make -d") too. I can do some such runs and attach the outputs as well.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 17:19:56](https://pagure.io/SSSD/sssd/issue/2350#226529)**

sh-4.2$ cat -n make_tests_02.log | grep libsss_nss_idmap.so
       617  libtool: link: gcc -shared  -fPIC -DPIC  src/sss_client/idmap/.libs/sss_nss_idmap.o src/sss_client/.libs/common.o src/util/.libs/strtonum.o   -lpthread -ldl  -O2   -Wl,-soname -Wl,libsss_nss_idmap.so.0 -o .libs/libsss_nss_idmap.so.0.0.1
       619  libtool: link: (cd ".libs" && rm -f "libsss_nss_idmap.so.0" && ln -s "libsss_nss_idmap.so.0.0.1" "libsss_nss_idmap.so.0")
       620  libtool: link: (cd ".libs" && rm -f "libsss_nss_idmap.so" && ln -s "libsss_nss_idmap.so.0.0.1" "libsss_nss_idmap.so")
    //libsss_nss_idmap.so was successfully linked.
    
      1750  libtool: link: rm -fr  .libs/libsss_nss_idmap.la .libs/libsss_nss_idmap.lai .libs/libsss_nss_idmap.so .libs/libsss_nss_idmap.so.0 .libs/libsss_nss_idmap.so.0.0.1
    // libsss_nss_idmap.so was removed. I don't know why
    
      1751  libtool: link: gcc -shared  -fPIC -DPIC  src/sss_client/idmap/.libs/sss_nss_idmap.o src/sss_client/.libs/common.o src/util/.libs/strtonum.o   -lpthread -ldl  -O2   -Wl,-soname -Wl,libsss_nss_idmap.so.0 -o .libs/libsss_nss_idmap.so.0.0.1
    //libsss_nss_idmap.so should be re-linked one more time
    //but another process wants to link with this library on line 1753
    
      1753  libtool: link: gcc -Wall -Wshadow -Wstrict-prototypes -Wpointer-arith -Wcast-qual -Wcast-align -Wwrite-strings -Wundef -Werror-implicit-function-declaration -Winit-self -fno-strict-aliasing -std=gnu99 -g -O2 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -o .libs/sss_nss_idmap-tests src/tests/cmocka/sss_nss_idmap_tests-sss_nss_idmap-tests.o  -lcmocka ./.libs/libsss_nss_idmap.so -lpthread -ldl
      1754  gcc: error: ./.libs/libsss_nss_idmap.so: No such file or directory

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 17:24:57](https://pagure.io/SSSD/sssd/issue/2350#226530)**

Are you sure that there are not two different builds stepping each other?

sssd in koji is build in parallel with 16 sub-processes
and there is not any problem.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 17:42:15](https://pagure.io/SSSD/sssd/issue/2350#226531)**

Removal could be a part of the linking rule.

Yes, I'm quite sure there is only one make running. Does koji build "tests" target as well?

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 18:12:08](https://pagure.io/SSSD/sssd/issue/2350#226532)**

removing symbolic links is part of linking rule (lines 619, 620) in comment 4. I do not see reason why 
file ".libs/libsss_nss_idmap.so.0.0.1" needs to be removed. I tried to run "make -j4 tests" on two different machines and was not able to reproduce problem. (file .libs/libsss_nss_idmap.so.0.0.1 was not removed after linking this library)
"make -j tests" passed as well.

koji runs command make twice: https://kojipkgs.fedoraproject.org//packages/sssd/1.12.0/2.fc21.beta1/data/logs/x86_64/build.log 
1. make -j4 all docs
2. make -j4 check

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-04 18:44:40](https://pagure.io/SSSD/sssd/issue/2350#226533)**

Did you try Fedora 20? I've seen this happen on Fedora 20 on a Brno Lab VM, and perhaps RHEL7 there as well. "make -j4 check" works fine for me too.

Could you try running, say, ten builds in a row?

However, this is not critical for me at the moment as I can use "make -j4 check LOG_COMPILER=true" instead.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-05 15:29:34](https://pagure.io/SSSD/sssd/issue/2350#226534)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.13 beta
rhbz:  => 0

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-05 16:33:16](https://pagure.io/SSSD/sssd/issue/2350#226535)**

Apparently this reproduces well when building on sshfs, probably because of the slowdown of file operations. However, I made it fail once on a native virtio drive in a Fedora 20 VM today. And I've seen failures on Brno Lab VMs as well, using native drives, which probably have slower IO operations there due to contention.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-05 16:46:52](https://pagure.io/SSSD/sssd/issue/2350#226536)**

I will try tomorrow building sssd in remote filesystem (sshfs)

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-05 21:50:09](https://pagure.io/SSSD/sssd/issue/2350#226537)**

This has reproduced in a Debian VM, building on sshfs as well.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 15:08:06](https://pagure.io/SSSD/sssd/issue/2350#226538)**

It has to be sshfs specific problem.

I tested build in VM with various file systems: **tmpfs, ext4, nfs** and it works for me without any probelm.

    sh-4.2$ file id_log.txt 
    id_log.txt: ERROR: cannot open `id_log.txt' (No such file or directory)
    sh-4.2$ for i in `seq 10`; do make -j4 tests && make -j4 clean && echo "$i" >> id_log.txt; done >/dev/null 2>&1
    
    sh-4.2$ 
    sh-4.2$ cat id_log.txt 
    1
    2
    3
    4
    5
    6
    7
    8
    9
    10
    sh-4.2$ cat /etc/issue
    Fedora release 20 (Heisenbug)
    Kernel \r on an \m (\l)

_comment0: It has to be sshfs specific problem.

I tested build in VM with various file systems: '''tmpfs, hdd, nfs''' and it works for me without any probelm.

{{{
sh-4.2$ file id_log.txt 
id_log.txt: ERROR: cannot open `id_log.txt' (No such file or directory)
sh-4.2$ for i in `seq 10`; do make -j4 tests && make -j4 clean && echo "$i" >> id_log.txt; done >/dev/null 2>&1

sh-4.2$ 
sh-4.2$ cat id_log.txt 
1
2
3
4
5
6
7
8
9
10
sh-4.2$ cat /etc/issue
Fedora release 20 (Heisenbug)
Kernel \r on an \m (\l)
}}} => 1402060360921978

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 15:15:13](https://pagure.io/SSSD/sssd/issue/2350#226539)**

The milestone of this ticket is set to the *SSSD 1.13 beta*. In my opinion, it is not problem in sssd. it must be problem with sshfs. We have other working alternatives. I would like to move this ticket to the milestone *SSSD Defered* or close as *NOTABUG*

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 15:56:27](https://pagure.io/SSSD/sssd/issue/2350#226540)**

I've seen this happen on other filesystems, although much less often. I haven't noticed any problems building other projects on sshfs, either. However, as I have a workaround, it is not critical. I wouldn't say it's NOTABUG, but rather "worksforme".

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 16:01:35](https://pagure.io/SSSD/sssd/issue/2350#226541)**

I cannot see any benefit to build on sshfs and I was not  able to reproduce it with nfs.

Would you prefer to move this ticket to *SSSD Defered* or close as *WORKSFORME*?

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 16:09:31](https://pagure.io/SSSD/sssd/issue/2350#226542)**

Let's close it. I'll reopen it once I get a better way to reproduce it, or it becomes a problem.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 16:16:51](https://pagure.io/SSSD/sssd/issue/2350#226543)**

It looks like you think it is a problem in sssd. There is a workaround. So, I will move this ticket to deferred
and it can be closed after few years :-)

Feel free to send patches if you find a solution.

milestone: SSSD 1.13 beta => SSSD Deferred

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-06 17:23:03](https://pagure.io/SSSD/sssd/issue/2350#226544)**

Log "make -j4 tests" failing on an xfs/virtio filesystem.
[rhel7_virtio_make_tests_failure.log.xz](/SSSD/sssd/issue/raw/files/5cffde9b3d48990aadca165f440c89f16b0cd244e45b072cc8bc97330678761f-rhel7_virtio_make_tests_failure.log.xz)

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-10 14:46:45](https://pagure.io/SSSD/sssd/issue/2350#226545)**

Important part of log rhel7_virtio_make_tests_failure.log.xz:

    libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -Wall -Iinclude -I..
      //snip
      -MT src/monitor/libsss_util_la-monitor_sbus.lo -MD -MP -MF
       src/monitor/.deps/libsss_util_la-monitor_sbus.Tpo 
       -c ../src/monitor/monitor_sbus.c  -fPIC -DPIC -o src/monitor/
       .libs/libsss_util_la-monitor_sbus.o
    
    /bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.  //snip
    /bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.  //snip
    mv -f src/monitor/.deps/libsss_util_la-monitor_sbus.Tpo src/monitor/.deps/libsss_util_la-monitor_sbus.Plo
    mv -f src/monitor/.deps/libsss_util_la-monitor_sbus.Tpo src/monitor/.deps/libsss_util_la-monitor_sbus.Plo
    mv: cannot stat ‘src/monitor/.deps/libsss_util_la-monitor_sbus.Tpo’: No such file or directory
    make[2]: *** [src/monitor/libsss_util_la-monitor_sbus.lo] Error 1
    make[2]: *** Waiting for unfinished jobs....

The command mv was executed twice and the second try was not successful. I am not sure why it was executed twice.

Is this issue reproducible without OOM in log files?

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-10 15:43:33](https://pagure.io/SSSD/sssd/issue/2350#226546)**

A process being killed seems very unlikely to trigger an additional execution of "mv". Still, the last log was taken on a system with 1GB of memory, which never had an OOM killer kill anything, according to journalctl.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-10 16:01:58](https://pagure.io/SSSD/sssd/issue/2350#226547)**

The last log was taken 4 days ago.

    
    Changed 4 days ago by nkondras
    
        Attachment rhel7_virtio_make_tests_failure.log.xz​ added
    
    Log "make -j4 tests" failing on an xfs/virtio filesystem.

The memory was increased yesterday. Did you mean another log file which is not attached in this ticket?

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-10 16:03:02](https://pagure.io/SSSD/sssd/issue/2350#226548)**

No, the last log was taken from a local VM running on my laptop.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2014-06-10 16:16:46](https://pagure.io/SSSD/sssd/issue/2350#226549)**

Replying to [comment:22 nkondras]:
> No, the last log was taken from a local VM running on my laptop. 

I see

Replying to [comment:20 nkondras]:
> A process being killed seems very unlikely to trigger an additional execution of "mv". Still, the last log was taken on a system with 1GB of memory, which never had an OOM killer kill anything, according to journalctl.
I cannot explain additional execution of "mv" either. It looks like another issue in automake or libtool.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2017-02-24 14:33:46](https://pagure.io/SSSD/sssd/issue/2350#226550)**

**Metadata Update from @nkondras**:
- Issue set to the milestone: SSSD Patches welcome

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2020-03-24 14:17:55](https://pagure.io/SSSD/sssd/issue/2350#636864)**

Thank you for taking time to submit this request for SSSD. Unfortunately this issue was not given priority and the team lacks the capacity to work on it at this time.

Given that we are unable to fulfill this request I am closing the issue as wontfix.

If the issue still persist on recent SSSD you can request re-consideration of this decision by reopening this issue. Please provide additional technical details about its importance to you.

Thank you for understanding.

---

**Comment from [nkondras](https://pagure.io/user/nkondras) at [2020-03-24 14:17:56](https://pagure.io/SSSD/sssd/issue/2350#636865)**

**Metadata Update from @pbrezina**:
- Issue close_status updated to: wontfix
- Issue status updated to: Closed (was: Open)
